app/src/main/java/org/mozilla/focus/topsites/DefaultTopSitesStorage.kt

```diff
@@ -0,0 +1,89 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.topsites
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import mozilla.components.concept.storage.FrecencyThresholdOption
+import mozilla.components.feature.top.sites.PinnedSiteStorage
+import mozilla.components.feature.top.sites.TopSite
+import mozilla.components.feature.top.sites.TopSitesStorage
+import mozilla.components.support.base.observer.Observable
+import mozilla.components.support.base.observer.ObserverRegistry
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Default implementation of [TopSitesStorage].
+ *
+ * @param pinnedSitesStorage An instance of [PinnedSiteStorage], used for storing pinned sites.
+ */
+class DefaultTopSitesStorage(
+    private val pinnedSitesStorage: PinnedSiteStorage,
+    coroutineContext: CoroutineContext = Dispatchers.IO
+) : TopSitesStorage, Observable<TopSitesStorage.Observer> by ObserverRegistry() {
+
+    private var scope = CoroutineScope(coroutineContext)
+
+    // Cache of the last retrieved top sites
+    var cachedTopSites = listOf<TopSite>()
+
+    override fun addTopSite(title: String, url: String, isDefault: Boolean) {
+        scope.launch {
+            pinnedSitesStorage.addPinnedSite(title, url, isDefault)
+
+            notifyObservers { onStorageUpdated() }
```

Comment:
(With Jetpack Compose this feature/subscriber mechanism seems super complicated and convoluted for now reason. Maybe that's something we could simplify eventually)